src/nunavut/lang/c/support/serialization.j2

```diff
@@ -0,0 +1,556 @@
+    return neg ? (int8_t)((-(int8_t)(uint8_t) ~val) - 1) : (int8_t) val;
+}
+
+static inline int16_t nunavutGetI16(const uint8_t* const           buf,
+                                    const size_t                   buf_size,
+                                    const {{typename_unsigned_bit_length}} off_bit,
+                                    const uint8_t                  len_bit)
+{
+    const uint8_t sat = (uint8_t) {{ choose_min('len_bit', '16U') }};
+    uint16_t      val = nunavutGetU16(buf, buf_size, off_bit, sat);
+    const bool    neg = (sat > 0U) && ((val & (1ULL << (sat - 1U))) != 0U);
+    val = ((sat < 16U) && neg) ? (uint16_t)(val | ~((1U << sat) - 1U)) : val;  // Sign extension
+    return neg ? (int16_t)((-(int16_t)(uint16_t) ~val) - 1) : (int16_t) val;
+}
+
+static inline int32_t nunavutGetI32(const uint8_t* const           buf,
+                                    const size_t                   buf_size,
+                                    const {{typename_unsigned_bit_length}} off_bit,
+                                    const uint8_t                  len_bit)
+{
+    const uint8_t sat = (uint8_t) {{ choose_min('len_bit', '32U') }};
+    uint32_t      val = nunavutGetU32(buf, buf_size, off_bit, sat);
+    const bool    neg = (sat > 0U) && ((val & (1ULL << (sat - 1U))) != 0U);
+    val = ((sat < 32U) && neg) ? (uint32_t)(val | ~((1UL << sat) - 1U)) : val;  // Sign extension
+    return neg ? (int32_t)((-(int32_t) ~val) - 1) : (int32_t) val;
+}
+
+static inline int64_t nunavutGetI64(const uint8_t* const           buf,
+                                    const size_t                   buf_size,
+                                    const {{typename_unsigned_bit_length}} off_bit,
+                                    const uint8_t                  len_bit)
+{
+    const uint8_t sat = (uint8_t) {{ choose_min('len_bit', '64U') }};
+    uint64_t      val = nunavutGetU64(buf, buf_size, off_bit, sat);
+    const bool    neg = (sat > 0U) && ((val & (1ULL << (sat - 1U))) != 0U);
+    val = ((sat < 64U) && neg) ? (uint64_t)(val | ~((1ULL << sat) - 1U)) : val;  // Sign extension
+    return neg ? (int64_t)((-(int64_t) ~val) - 1) : (int64_t) val;
+}
+
+// --------------------------------------------- PUBLIC API - FLOAT16 ---------------------------------------------
+
+{%- if not option_omit_float_serialization_support %}
+
+static_assert(NUNAVUT_PLATFORM_IEEE754_FLOAT, "The target platform does not support IEEE754 floating point operations.");
+static_assert(32U == (sizeof({{typename_float_32}}) * 8U), "Unsupported floating point model");
+
+// Intentional violation of MISRA: we need this union because the alternative is far more error prone.
+// We have to rely on low-level data representation details to do the conversion; unions are helpful.
+typedef union  // NOSONAR
+{
+    uint32_t       bits;
+    {{typename_float_32}} real;
+} Float32Bits;
```

Comment:
Please define this inside of the functions that use the union internally. Again, be careful not to pollute the global namespace. Use a Jinja macro if you want.